ladybug/types/_base.py

```diff
@@ -0,0 +1,226 @@
+        )
+
+    def to_ip(self, values, from_unit=None):
+        """Converts a list of values to IP from a given from_unit."""
+        raise NotImplementedError(
+            'to_ip is not implemented on %s' % self.__class__.__name__
+        )
+
+    def to_si(self, values, from_unit=None):
+        """Convert a list of values to SI from a given from_unit."""
+        raise NotImplementedError(
+            'to_si is not implemented on %s' % self.__class__.__name__
+        )
+
+    def is_missing(self, value):
+        """Check if a value contains missing data when in an EPW."""
+        if value == self.missing_epw:
+            return True
+        return False
+
+    def is_in_range(self, values, unit=None, raise_exception=False):
+        """Check if a list of values is within acceptable ranges.
+
+        Args:
+            values: A list of values.
+            unit: The unit of the values.  If not specified, the default metric
+                unit will be assumed.
+            raise_exception: Set to True to raise an exception if not in range.
+        """
+        self._check_values(values)
+        if unit is None or unit == self.units[0]:
+            minimum = self.min
+            maximum = self.max
+        else:
+            namespace = {'self': self, 'minimum': None, 'maximum': None}
+            self.is_unit_acceptable(unit, True)
+            min_statement = "minimum = self._{}_to_{}(self.min)".format(
+                self._clean(self.units[0]), self._clean(unit))
+            max_statement = "maximum = self._{}_to_{}(self.max)".format(
+                self._clean(self.units[0]), self._clean(unit))
+            exec(min_statement, namespace)
+            exec(max_statement, namespace)
+            minimum, maximum = namespace['minimum'], namespace['maximum']
+
+        for value in values:
+            if value < minimum or value > maximum:
+                if not raise_exception:
+                    return False
+                else:
+                    raise ValueError(
+                        '{0} should be between {1} and {2}. Got {3}'.format(
+                            self.__class__.__name__, self.min, self.max, value
+                        )
+                    )
+        return True
+
+    def is_in_range_epw(self, values, unit=None, raise_exception=False):
+        """Check if a list of values is within acceptable ranges for an EPW file.
+
+        Args:
+            values: A list of values.
+            unit: The unit of the values.  If not specified, the default metric
+                unit will be assumed.
+            raise_exception: Set to True to raise an exception if not in range.
+        """
+        self._check_values(values)
+        if unit is None or unit == self.units[0]:
+            minimum = self.min_epw
+            maximum = self.max_epw
+        else:
+            namespace = {'self': self, 'minimum': None, 'maximum': None}
```

Comment:
There may be some convoluted ways to avoid some of them.
My question to you is, why do we need to avoid `exec`?
Maybe if I understand that, I can understand if some of them should get replaced.

Reply:
I am taking out all execs now.